File: packages/hht_auth/hht_auth_server/lib/src/repositories/user_repository.dart
/// IMPLEMENTS REQUIREMENTS:
///   REQ-d00081: User Document Schema - Repository interface for user data access
///
/// Abstract repository interface for user data access.

import 'package:hht_auth_core/hht_auth_core.dart';

/// Repository interface for WebUser persistence operations.
abstract class UserRepository {
  /// Creates a new user in the database.
  ///
  /// Throws [AuthException] if username already exists for the sponsor.
  Future<void> createUser(WebUser user);

  /// Retrieves a user by username and sponsor ID.
  ///
  /// Returns null if user not found.
  Future<WebUser?> getUserByUsername(String username, String sponsorId);

  /// Retrieves a user by user ID.
  ///
  /// Returns null if user not found.
  Future<WebUser?> getUserById(String userId);

  /// Updates an existing user.
  ///
  /// Throws [AuthException] if user not found.
  Future<void> updateUser(WebUser user);

  /// Increments the failed login attempts counter.
  ///
  /// Returns the updated user.
  Future<WebUser> incrementFailedAttempts(String userId);

  /// Resets the failed login attempts counter and clears lock.
  ///
  /// Returns the updated user.
  Future<WebUser> resetFailedAttempts(String userId);

  /// Locks a user account until the specified time.
  ///
  /// Returns the updated user.
  Future<WebUser> lockAccount(String userId, DateTime lockedUntil);
}

File: packages/hht_auth/hht_auth_server_ts/src/repositories/user.repository.ts
/**
 * IMPLEMENTS REQUIREMENTS:
 *   REQ-d00081: User Document Schema
 *
 * User repository interface and in-memory implementation.
 */

import { WebUser } from '../models/web-user.model.js';

/**
 * Repository interface for user operations.
 */
export interface UserRepository {
  createUser(user: WebUser): Promise<void>;
  getUserByUsername(username: string): Promise<WebUser | null>;
  getUserById(userId: string): Promise<WebUser | null>;
  updateUser(user: WebUser): Promise<void>;
  usernameExists(username: string): Promise<boolean>;
}

/**
 * In-memory user repository for testing.
 */
export class InMemoryUserRepository implements UserRepository {
  private users: Map<string, WebUser> = new Map();

  async createUser(user: WebUser): Promise<void> {
    this.users.set(user.id, { ...user });
  }

  async getUserByUsername(username: string): Promise<WebUser | null> {
    for (const user of this.users.values()) {
      if (user.username.toLowerCase() === username.toLowerCase()) {
        return { ...user };
      }
    }
    return null;
  }

  async getUserById(userId: string): Promise<WebUser | null> {
    const user = this.users.get(userId);
    return user ? { ...user } : null;
  }

  async updateUser(user: WebUser): Promise<void> {
    if (!this.users.has(user.id)) {
      throw new Error(`User ${user.id} not found`);
    }
    this.users.set(user.id, { ...user });
  }

  async usernameExists(username: string): Promise<boolean> {
    for (const user of this.users.values()) {
      if (user.username.toLowerCase() === username.toLowerCase()) {
        return true;
      }
    }
    return false;
  }

  /** Clear all users (for testing) */
  clear(): void {
    this.users.clear();
  }

  /** Seed users (for testing) */
  seed(users: WebUser[]): void {
    for (const user of users) {
      this.users.set(user.id, { ...user });
    }
  }
}
